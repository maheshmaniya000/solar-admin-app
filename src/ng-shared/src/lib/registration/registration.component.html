<button
  class="sg-small-button"
  mat-flat-button
  color="accent"
  (click)="onOpenRegistrationButtonClick()"
  data-test="open-registration-button"
>
  {{ 'auth.registration.title' | transloco }}
</button>

<ng-template #registrationDialogContent>
  <ng-container
    *ngIf="registrationStore.registrationDialogViewModel$ | async as viewModel"
  >
    <sg-dialog
      *transloco="let t; read: 'auth.registration'"
      [title]="t('createUserAccount')"
      [closeIconButtonEnabled]="true"
      [dividedContent]="false"
    >
      <mat-horizontal-stepper
        linear
        class="no-header"
        [selectedIndex]="viewModel.step"
      >
        <mat-step> step1 </mat-step>
        <mat-step> step2 </mat-step>
        <mat-step> step3 </mat-step>
        <mat-step> step4 </mat-step>
        <mat-step> step5 </mat-step>
        <mat-step> step6 </mat-step>
      </mat-horizontal-stepper>

      <mat-dialog-actions>
        <button
          *ngIf="viewModel.backButtonDisplayed"
          mat-button
          color="primary"
          (click)="registrationStore.setPreviousStep()"
          data-test="back-button"
        >
          <mat-icon>arrow_back</mat-icon>
          {{ 'common.action.back' | transloco }}
        </button>
        <span class="spacer"></span>
        <button
          class="sg-large-button"
          *ngIf="viewModel.continueButtonDisplayed"
          mat-flat-button
          color="accent"
          (click)="registrationStore.setNextStep()"
          data-test="continue-button"
        >
          {{ 'common.action.continue' | transloco }}
        </button>
      </mat-dialog-actions>

      <sg-dialog-footer *ngIf="viewModel.footerDisplayed">
        {{ t('alreadyHaveAccount') }}
        <button
          mat-button
          color="accent"
          (click)="onLogInButtonClick()"
          data-test="login-button"
        >
          {{ t('login') }}
        </button>
      </sg-dialog-footer>
    </sg-dialog>
  </ng-container>
</ng-template>
