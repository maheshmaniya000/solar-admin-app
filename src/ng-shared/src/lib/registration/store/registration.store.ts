import { Injectable } from '@angular/core';
import { ComponentStore } from '@ngrx/component-store';
import { Observable } from 'rxjs';

import { AccountType } from '../models/account-type.enum';
import { RegistrationStep } from '../models/registration-step.enum';

type RegistrationState = {
  accountType: AccountType;
  step: RegistrationStep;
};

type RegistrationViewModel = {
  step: RegistrationStep;
  backButtonDisplayed: boolean;
  continueButtonDisplayed: boolean;
  footerDisplayed: boolean;
};

const initialState: RegistrationState = {
  step: RegistrationStep.selectAccountType,
  accountType: AccountType.work
};

@Injectable()
export class RegistrationStore extends ComponentStore<RegistrationState> {
  readonly setPreviousStep = this.updater(state => ({
    ...state,
    step: --state.step
  }));

  readonly setNextStep = this.updater(state => ({
    ...state,
    step: ++state.step
  }));

  readonly setAccountType = this.updater((state, accountType: AccountType) => ({
    ...state,
    accountType
  }));

  readonly registrationDialogViewModel$: Observable<RegistrationViewModel> =
    this.select(state => ({
      ...state,
      backButtonDisplayed: state.step !== RegistrationStep.selectAccountType,
      continueButtonDisplayed:
        state.step !== RegistrationStep.accountVerification,
      footerDisplayed: state.step === RegistrationStep.selectAccountType
    }));

  constructor() {
    super(initialState);
  }
}
