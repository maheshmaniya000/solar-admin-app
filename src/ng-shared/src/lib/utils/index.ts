export * from './router/custom.reuse-stragegy';
export * from './router/custom.router-state.serializer';
export * from './router/custom.url-serializer';
export * from './rxjs/simplechange.operator';
export * from './app-device.utils';
export * from './dom.utils';
export * from './enum.utils';
export * from './misc';
export * from './module.utils';
export * from './response';
export * from './timezone.utils';
export * from './translation.utils';
export * from './validators';
