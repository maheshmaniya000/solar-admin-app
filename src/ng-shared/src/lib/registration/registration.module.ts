import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { TranslocoModule } from '@ngneat/transloco';

import { DialogModule } from 'components/dialog/dialog.module';
import { StepperModule } from 'components/stepper/stepper.module';

import { RegistrationComponent } from './registration.component';

@NgModule({
  imports: [
    CommonModule,
    MatButtonModule,
    MatIconModule,
    TranslocoModule,
    DialogModule,
    StepperModule
  ],
  declarations: [RegistrationComponent],
  exports: [RegistrationComponent]
})
export class RegistrationModule {}
