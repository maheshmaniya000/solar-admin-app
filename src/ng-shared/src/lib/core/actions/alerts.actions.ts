import { Action } from '@ngrx/store';

import { ProjectId } from '@solargis/types/project';

import { Alert } from '../types';


export const ALERT_ADD = '[alert] add';
export const ALERT_CLOSE = '[alert] close';
export const ALERT_PROJECT_CLOSE = '[alert] project close';

export class AddAlert implements Action {
  readonly type = ALERT_ADD;
  constructor(public alert: Alert) {}
}

export class CloseAlerts implements Action {
  readonly type = ALERT_CLOSE;
  constructor(public alerts: Alert[]) {}
}

export class CloseProjectAlerts implements Action {
  readonly type = ALERT_PROJECT_CLOSE;
  constructor(public projectId: ProjectId ) {}
}

export type AlertActions = AddAlert | CloseAlerts | CloseProjectAlerts;
