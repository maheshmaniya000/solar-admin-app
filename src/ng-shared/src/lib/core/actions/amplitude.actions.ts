import { Action } from '@ngrx/store';

import { AmplitudeEventCode, AmplitudeEventData } from '../services/amplitude-analytics.service';

export const AMPLITUDE_TRACK_EVENT = '[amplitude] track event';

export class AmplitudeTrackEvent implements Action {
  readonly type = AMPLITUDE_TRACK_EVENT;

  constructor(public event: AmplitudeEventCode, public data: AmplitudeEventData = {}) {}
}
