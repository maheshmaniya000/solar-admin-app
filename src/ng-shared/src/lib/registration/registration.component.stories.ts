import { APP_BASE_HREF } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { provideMockStore } from '@ngrx/store/testing';
import { moduleMetadata } from '@storybook/angular';

import { SgDefaultsModule } from 'components/defaults/defaults.module';

import { TranslocoRootModule } from '../core/transloco-root.module';
import { RegistrationModule } from './registration.module';

export default {
  title: 'Solargis/Components/Registration',
  decorators: [
    moduleMetadata({
      imports: [
        BrowserAnimationsModule,
        SgDefaultsModule,
        RouterTestingModule.withRoutes([]),
        HttpClientModule,
        TranslocoRootModule,
        RegistrationModule
      ],
      providers: [{ provide: APP_BASE_HREF, useValue: '.' }, provideMockStore()]
    })
  ]
};

export const basic = (props): any => ({
  props,
  template: `
    <sg-registration></sg-registration>
  `
});
