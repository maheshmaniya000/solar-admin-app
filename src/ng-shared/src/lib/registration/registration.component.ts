import {
  ChangeDetectionStrategy,
  Component,
  TemplateRef,
  ViewChild
} from '@angular/core';
import { Store } from '@ngrx/store';

import { DialogService } from 'components/dialog/service/dialog.service';

import { RequireUserLogin } from '../user/actions/auth.actions';
import { RegistrationStep } from './models/registration-step.enum';
import { RegistrationStore } from './store/registration.store';

@Component({
  selector: 'sg-registration',
  templateUrl: './registration.component.html',
  providers: [RegistrationStore],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class RegistrationComponent {
  @ViewChild('registrationDialogContent')
  registrationDialogContent: TemplateRef<any>;

  readonly registrationStep = RegistrationStep;

  constructor(
    readonly registrationStore: RegistrationStore,
    private readonly dialogService: DialogService,
    private readonly store: Store
  ) {}

  onOpenRegistrationButtonClick(): void {
    this.dialogService.openLargeDialog(this.registrationDialogContent, {
      disableClose: true
    });
  }

  onLogInButtonClick(): void {
    this.store.dispatch(new RequireUserLogin(undefined));
  }
}
