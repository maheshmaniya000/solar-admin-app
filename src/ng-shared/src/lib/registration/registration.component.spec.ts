import { HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { MatDialogModule } from '@angular/material/dialog';
import {
  createComponentFactory,
  createSpyObject,
  Spectator,
  SpyObject
} from '@ngneat/spectator/jest';
import { TranslocoTestingModule } from '@ngneat/transloco';
import { Store } from '@ngrx/store';

import { DialogService } from 'components/dialog/service/dialog.service';
import { expectToMatchMatDialogSnapshot } from 'components/utils/test';
import { toDataTestAttributeSelector } from 'components/utils/test/to-data-test-attribute-selector';
import { RequireUserLogin } from 'ng-shared/user/actions/auth.actions';

import { RegistrationStep } from './models/registration-step.enum';
import { RegistrationComponent } from './registration.component';
import { RegistrationStore } from './store/registration.store';

import spyOn = jest.spyOn;

xdescribe('RegistrationComponent', () => {
  describe('unit', () => {
    let component: RegistrationComponent;
    let dialogServiceSpyObject: SpyObject<DialogService>;
    let storeSpyObject: SpyObject<Store>;

    beforeEach(() => {
      storeSpyObject = createSpyObject(Store);
      dialogServiceSpyObject = createSpyObject(DialogService);
      component = new RegistrationComponent(
        undefined,
        dialogServiceSpyObject,
        storeSpyObject
      );
    });

    describe('onOpenRegistrationButtonClick', () => {
      it('should open large dialog with provided template', () => {
        component.onOpenRegistrationButtonClick();
        expect(dialogServiceSpyObject.openLargeDialog).toHaveBeenCalledWith(
          component.registrationDialogContent,
          {
            disableClose: true
          }
        );
      });

      describe('onLogInButtonClick', () => {
        it('should dispatch require login action', () => {
          component.onLogInButtonClick();
          expect(storeSpyObject.dispatch).toHaveBeenCalledTimes(1);
          expect(storeSpyObject.dispatch).toHaveBeenCalledWith(
            new RequireUserLogin(undefined)
          );
        });
      });
    });
  });

  describe('component', () => {
    let spectator: Spectator<RegistrationComponent>;
    const createComponent = createComponentFactory({
      component: RegistrationComponent,
      imports: [MatDialogModule, TranslocoTestingModule],
      providers: [DialogService, RegistrationStore],
      mocks: [Store],
      shallow: true
    });

    const clickOnOpenRegistrationButton = (): void =>
      spectator.click(toDataTestAttributeSelector('open-registration-button'));

    beforeEach(() => {
      spectator = createComponent();
    });

    it(`should match snapshot with no dialog and have correctly bound
 methods`, () => {
      const onOpenRegistrationButtonClickSpy = spyOn(
        spectator.component,
        'onOpenRegistrationButtonClick'
      );
      clickOnOpenRegistrationButton();
      expect(onOpenRegistrationButtonClickSpy).toHaveBeenCalledTimes(1);
      expect(spectator.fixture).toMatchSnapshot();
    });

    const continueButtonSelector =
      toDataTestAttributeSelector('continue-button');
    const loginButtonSelector = toDataTestAttributeSelector('login-button');

    describe('onOpenRegistrationButtonClick', () => {
      let rootLoader: HarnessLoader;

      beforeEach(() => {
        rootLoader = TestbedHarnessEnvironment.documentRootLoader(
          spectator.fixture
        );
      });

      it(`should match snapshot of dialog with continue button, footer,
 without back button and have correctly bound attributes and methods`, () => {
        clickOnOpenRegistrationButton();
        expect(
          spectator.query('mat-horizontal-stepper', { root: true })
        ).toEqual(
          expect.objectContaining({
            selectedIndex: 0
          })
        );
        expect(spectator.query('sg-dialog', { root: true })).toEqual(
          expect.objectContaining({
            closeIconButtonEnabled: true,
            dividedContent: false
          })
        );

        const setNextStepSpy = spyOn(
          spectator.component.registrationStore,
          'setNextStep'
        );
        spectator.click(
          spectator.query(continueButtonSelector, {
            root: true
          })
        );
        expect(setNextStepSpy).toHaveBeenCalledTimes(1);

        const onLoginButtonClickSpy = spyOn(
          spectator.component,
          'onLogInButtonClick'
        );
        spectator.click(
          spectator.query(loginButtonSelector, {
            root: true
          })
        );
        expect(onLoginButtonClickSpy).toHaveBeenCalledTimes(1);

        expectToMatchMatDialogSnapshot(rootLoader);
      });

      it(`should contain back button with correctly bound method and not
 contain continue button and login button`, () => {
        spectator.component.registrationStore.patchState({
          step: RegistrationStep.accountVerification
        });
        clickOnOpenRegistrationButton();

        const setPreviousStepSpy = spyOn(
          spectator.component.registrationStore,
          'setPreviousStep'
        );
        spectator.click(
          spectator.query(toDataTestAttributeSelector('back-button'), {
            root: true
          })
        );
        expect(setPreviousStepSpy).toHaveBeenCalledTimes(1);

        expect(
          spectator.query(continueButtonSelector, { root: true })
        ).not.toExist();

        expect(
          spectator.query(loginButtonSelector, { root: true })
        ).not.toExist();
      });
    });
  });
});
