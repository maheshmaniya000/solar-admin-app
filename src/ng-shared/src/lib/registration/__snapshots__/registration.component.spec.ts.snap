// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationComponent component onOpenRegistrationButtonClick should match snapshot of dialog with continue button, footer,
 without back button and have correctly bound attributes and methods 1`] = `
<mat-dialog-container
  aria-modal="true"
  class="mat-dialog-container ng-trigger ng-trigger-dialogContainer"
  role="dialog"
  style="transform:none;0:transform;"
  tabindex="-1"
>
  <sg-dialog
    style=""
    title="en.auth.registration.createUserAccount"
  >
    <mat-horizontal-stepper
      class="no-header"
      linear=""
    >
      <mat-step>
         step1
      </mat-step>
      <mat-step>
         step2
      </mat-step>
      <mat-step>
         step3
      </mat-step>
      <mat-step>
         step4
      </mat-step>
      <mat-step>
         step5
      </mat-step>
      <mat-step>
         step6
      </mat-step>
    </mat-horizontal-stepper>
    <mat-dialog-actions
      class="mat-dialog-actions"
    >

      <span
        class="spacer"
      />
      <button
        class="sg-large-button"
        color="accent"
        data-test="continue-button"
        mat-flat-button=""
      >
         en.common.action.continue
      </button>

    </mat-dialog-actions>
    <sg-dialog-footer>
       en.auth.registration.alreadyHaveAccount
      <button
        color="accent"
        data-test="login-button"
        mat-button=""
      >
         en.auth.registration.login
      </button>
    </sg-dialog-footer>

  </sg-dialog>




</mat-dialog-container>
`;

exports[`RegistrationComponent component should match snapshot with no dialog and have correctly bound
 methods 1`] = `
<sg-registration
  RegistrationStep={[Function Object]}
  __ngContext__={[Function LRootView]}
  dialogService={[Function DialogService]}
  onOpenRegistrationButtonClick={[Function Function]}
  registrationDialogContent={[Function TemplateRef]}
  registrationStore={[Function RegistrationStore]}
  store={[Function Object]}
>
  <button
    class="sg-small-button"
    color="accent"
    data-test="open-registration-button"
    mat-flat-button=""
  >
     en.auth.registration.title
  </button>
</sg-registration>
`;
